db/selector.go

```diff
@@ -0,0 +1,145 @@
+package db
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+type typeSelector[T any] struct {
+	wrapped MappedExecutor
+}
+
+// NewTypeSelector returns an object which can be used to automagically query
+// the provided type-mapped database for rows of the parameterized type.
+func NewTypeSelector[T any](wm MappedExecutor) *typeSelector[T] {
+	return &typeSelector[T]{wrapped: wm}
+}
+
+// SelectRows combines the best features of gorp, the go stdlib, and generics.
+// It uses the type parameter of the typeSelector object to automatically look
+// up the proper table name and columns to select. It returns an iterable which
+// yields fully-populated objects of the parameterized type directly. The given
+// clauses MUST be only the bits of a sql query from "WHERE ..." onwards; if
+// they contain any of the "SELECT ... FROM ..." portion of the query it will
+// result in an error. The caller is responsible for calling `Rows.Close()`
+// when they are done with the query.
+func (ts typeSelector[T]) SelectRows(ctx context.Context, clauses string, args ...interface{}) (*typeRows[T], error) {
+	// Look up the table to use based on the type of this TypeSelector.
+	var t T
+	tableMap, err := ts.wrapped.TableFor(reflect.TypeOf(t), false)
+	if err != nil {
+		return nil, fmt.Errorf("database model type not mapped to table name: %w", err)
+	}
+
+	return ts.SelectRowsFrom(ctx, tableMap.TableName, clauses, args...)
+}
+
+// SelectRowsFrom is the same as SelectRows, but it additionally takes a table
+// name to select from, rather than automatically computing the table name from
+// gorp's DbMap.
+func (ts typeSelector[T]) SelectRowsFrom(ctx context.Context, tablename string, clauses string, args ...interface{}) (*typeRows[T], error) {
+	// Look up the table to use based on the type of this TypeSelector.
+	var throwaway T
+	t := reflect.TypeOf(throwaway)
+	tableMap, err := ts.wrapped.TableFor(t, false)
+	if err != nil {
+		return nil, fmt.Errorf("database model type not mapped to table name: %w", err)
+	}
+
+	// Extract the list of column names from the type's struct tags.
```

Comment:
This comment is a little off; it takes the list of column names from the tableMap (which, I think, got them from the type's struct tags when it was built?)